### ISSUE TITLE
Poor performance with rayon ParallelIterator

### ISSUE THREAD

**Issue by pkolaczk** [console-rs/indicatif#170]
When processing several hundreds of thousands of items per second with a `ParallelIterator`, `ProgressBar` becomes a bottleneck because it grabs a global shared lock whenever the state needs to be updated. Instead, it would be better to store internal progress for each rayon thread separately and sum these counters on read. 

Having said that as a Rust beginner, I'm not sure how hard such implementation could be in Rust and whether Rust has a similar structure like Java `LongAdder` (https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/LongAdder.html).

`RelaxedCounter` from https://docs.rs/atomic-counter/1.0.1/atomic_counter/ maybe?

**Comment by mibac138** [comment]
By global shared lock do you mean the `stderr` lock? If writing to the console is your bottleneck, you can use `ProgressBar::set_draw_delta` to limit the number of updates. The bigger the number you provide the less frequent the updates are going to be and thus the less locking will occur.

**Comment by pkolaczk** [comment]
No, stderr is not a problem. I already tried reducing the rendering refresh rate and it didn't help at all.

The problem is the fact that the whole `ProgressBar` is wrapped in an `RwLock` :

See this in `rayon_support`:
```rust 
    pub struct ParProgressBarIter<T> {
        it: T,
        progress: Arc<Mutex<ProgressBar>>,
    }
```

And also updating the internal state requires locking:
```
    fn update_and_draw<F: FnOnce(&mut ProgressState)>(&self, f: F) {
        let mut draw = false;
        {
            let mut state = self.state.write().unwrap();
```

**Comment by pkolaczk** [comment]
As a quick check I tried using `RelaxedCounter` and it seems to work much better on my iterator that delivers about ~1 million items / second.

Here a fragment of my original code:

```
    let file_scan_pb = ProgressBar::new_spinner();
    file_scan_pb.set_draw_target(ProgressDrawTarget::stdout_with_hz(1));
    file_scan_pb.set_style(ProgressStyle::default_spinner().template("Scanning files: {pos}"));

    let files = ... // obtain a parallel iterator
    files
        .progress_with(file_scan_pb)
        .foreach(|item| ...);  
```

And after changing to the counter:

```rust
    let counter = RelaxedCounter::new(0);
    let files = ... // obtain a parallel iterator   
    files
        .inspect(|_| { counter.inc(); () })
        .foreach(|item| ...);  
```

This one is about 3x faster, just as fast as without the `inspect()` line at all.

**Comment by mibac138** [comment]
Oh, I see the problem now. Could you try and use an experimental branch to see if it helps?
To do that you will need to modify your `Cargo.toml`. Change `indicatif = "0.14.0"` to `indicatif = { git="https://github.com/mibac138/indicatif", branch="rayon-rate-limit" }`. Then try it out like so:
```rust
let pb = ...
let iter = pb.wrap_par_iter(files);
iter.set_update_delta(1000);
iter.foreach(|item| ...);
```

**Comment by pkolaczk** [comment]
```
error[E0599]: no method named `wrap_par_iter` found for struct `indicatif::progress::ProgressBar` in the current scope
  --> src/main.rs:62:36
   |
62 |     let size_groups = file_scan_pb.wrap_par_iter(files)
   |                                    ^^^^^^^^^^^^^ help: there is a method with a similar name: `wrap_iter`

```

**Comment by mibac138** [comment]
Oops, please make sure you also enable `with_rayon` feature like so `indicatif = { git="https://github.com/mibac138/indicatif", branch="rayon-rate-limit, features = ["with_rayon"] }`

**Comment by pkolaczk** [comment]
Does not compile :(
```
   Compiling indicatif v0.14.0 (https://github.com/mibac138/indicatif?branch=rayon-rate-limit#555b0eb6)
error[E0425]: cannot find value `update_delta` in this scope
  --> /home/pkolaczk/.cargo/git/checkouts/indicatif-4be6e86cdcc437b6/555b0eb/src/par_iter.rs:70:36
   |
70 |         self.data.set_update_delta(update_delta);
   |                                    ^^^^^^^^^^^^ not found in this scope

error[E0609]: no field `progress` on type `&par_iter::ProgressConsumer<C>`
   --> /home/pkolaczk/.cargo/git/checkouts/indicatif-4be6e86cdcc437b6/555b0eb/src/par_iter.rs:119:64
    |
119 |         ProgressConsumer::new(self.base.split_off_left(), self.progress.clone())
    |                                                                ^^^^^^^^ unknown field
    |
    = note: available fields are: `base`, `data`

error[E0615]: attempted to take value of method `progress` on type `par_iter::ParProgressBarIter<T>`
   --> /home/pkolaczk/.cargo/git/checkouts/indicatif-4be6e86cdcc437b6/555b0eb/src/par_iter.rs:156:62
    |
156 |         let consumer1 = ProgressConsumer::new(consumer, self.progress);
    |                                                              ^^^^^^^^ help: use parentheses to call the method: `progress()`

error: aborting due to 3 previous errors

```

**Comment by mibac138** [comment]
It looks like the errors are in indicatif, however I have already fixed them. Running `cargo update` should make it work.

**Comment by pkolaczk** [comment]
Looks like it works, however I had to disable the refresh rate, because otherwise it didn't write anything at all. 

update_delta = 1:
```
real    0m9,160s
user    0m11,640s
sys     0m12,154s
```

update_delta = 10:
```
real    0m3,252s
user    0m7,111s
sys     0m8,080s
```


update_delta=100:
```
real    0m2,612s
user    0m6,521s
sys     0m8,098s
```

No progressbar at all:
```
real    0m2,535s
user    0m6,023s
sys     0m7,870s
```
I'm not entirely happy with this workaround, though, because the "speed" at which my iterator is able to deliver items is heavily dependent on mny circumstances. I'd rather progress bar recorded all updates fast like in the example with the `RelaxedCounter`. 
{noformat}

**Comment by mibac138** [comment]
> BTW: Is it possible to do a pull-based progress bar? Like - I could provide a RelaxedCounter and PB would only fetch its value and display repeatedly?

You could write your own `enable_steady_tick` implementation and it should work. Something like this:
```rust
let counter = Arc::new(counter); // use this in inspect(counter.inc())

let counter2 = Arc::clone(&counter);
let pb2 = pb.clone();
thread::spawn(move || {
    if Arc::strong_count(&counter2) == 1 {
        // intensive counting is over, stop updating in a separate thread
        break;
    }
    thread::sleep(interval);
    pb2.set_position(counter2.get());
});
```

**Comment by pkolaczk** [comment]
Thank you! 

This is nice:
```
if Arc::strong_count(&counter2) == 1 {
        // intensive counting is over, stop updating in a separate thread
        break;
    }
```
I believe this could be also made to hide the progress bar once the counter is destroyed. :)

**Comment by mibac138** [comment]
> I believe this hides the progress bar once the counter is dropped? Is it right?

TLDR: Yes

When a ProgressBar is dropped it is hidden unless you have called `finish` or `finish_with_message`. The problem is that for the progress bar to be dropped all `Arc`s holding it must be dropped too. This check makes sure that the `Arc` owned by the spawned thread is dropped so that the `ProgressBar` can be dropped too. It could be made simpler but it requires #161 which isn't yet merged.

**Comment by thomwiggers** [comment]
Using @mibac138 's fork gives me a lot of 

```
thread '<unnamed>' panicked at 'attempt to divide by zero', /home/thom/.cargo/git/checkouts/indicatif-4be6e86cdcc437b6/c716edb/src/par_iter.rs:47:26
```

:(

**Comment by sstadick** [comment]
Is this a use case that `indicatif` is intending to support as a first class option? I also ran into this issue (indicatif being the bottleneck in a hot `par_iter` loop and am looking for fixes.

I'm also willing to making a PR to fix/enable this with something along the `enable_steady_tick` route if you are open to that

**Comment by Ekleog** [comment]
FWIW, this problem is also very visible outside of rayon: I'm parsing gzipped log files that have a few million lines, and depending on the option I pick I get vastly different speeds:
- If I never `inc` the progress bar, it takes 34 seconds
- If I `.wrap_read()` the file (before giving it to `libflate::gzip::Decoder` because otherwise it's going way over 100%), it takes about 20 minutes (never actually tried waiting that long, but it's the indicatif-provided estimate)

And when parsing another, non-gzipped log file:
- If I never `inc` the bar, it completes in 2 seconds
- If I `inc` the bar by blocks of `file size / 2048`, it completes in 2 seconds
- If I `.wrap_read()` the file, it completes in 4 seconds
- If I `inc` the bar on each line read, it completes in 1 min 5 sec

Overall, I'd think that the best option might be to “just” make ProgressBar's internal counter an AtomicU64, and then instead of precomputing all the metadata on each `inc()` (which naturally requires a mutex), to re-compute it on each tick. This being said, I'm far from familiar with the internals of `indicatif`, so… this may be impossible for some reason I don't see? But if it is possible, I think it'd solve all the issues at the same time, while also providing the best user interface.

WDYT?

**Comment by untitaker** [comment]
If you're reading this thread because the progressbar is slowing down your application and don't yet know exactly why, try using `set_draw_rate()` and check if it still does. In my usage of indicatif I found that the application is primarily bottlenecked on terminal output, not any of what OP is talking about, even when processing 10k items per second. See also #363

**Comment by djc** [comment]
If someone can retest this against current main and still sees bad performance, I'd like to hear about it. The internal locking is still pretty much the same, but a bunch of other improvements should have improved performance.

**Comment by mooreniemi** [comment]
In recent versions `set_draw_delta` is gone and only `enable_steady_tick` remains. Can we get maintainer guidance on the performance impact?

**Comment by mooreniemi** [comment]
```
pb.enable_steady_tick(Duration::from_secs(30));
```

Draws less often to screen, but based on time. It appears now there's not a way to minimize draws based on number of input elements (what we could do with `set_draw_delta`).

0.16.2:

```
    /// Limit redrawing of progress bar to every `n` steps
    ///
    /// By default, the progress bar will redraw whenever its state advances. This setting is
    /// helpful in situations where the overhead of redrawing the progress bar dominates the
    /// computation whose progress is being reported.
    ///
    /// If `n` is greater than 0, operations that change the progress bar such as
    /// [`ProgressBar::tick()`], [`ProgressBar::set_message()`] and [`ProgressBar::set_length()`]
    /// will no longer cause the progress bar to be redrawn, and will only be shown once the
    /// position advances by `n` steps.
    ///
    /// ```rust,no_run
    /// # use indicatif::ProgressBar;
    /// let n = 1_000_000;
    /// let pb = ProgressBar::new(n);
    /// pb.set_draw_delta(n / 100); // redraw every 1% of additional progress
    /// ```
    ///
    /// Note that `ProgressDrawTarget` may impose additional buffering of redraws.
    pub fn set_draw_delta(&self, n: u64) {
        let mut state = self.state.lock().unwrap();
        state.draw_delta = n;
        state.draw_next = state.pos.saturating_add(state.draw_delta);
    }
```

Haven't read lib code deeply yet but my understanding is these choices have a performance impact.

**Comment by mooreniemi** [comment]
For `set_draw_delta` that I see above there, seems obvious we'd want to avoid a `lock` in favor of some type of atomic counter (as mentioned above). But haven't dug into the mechanics of the timed ticker.

**Comment by mooreniemi** [comment]
[Just to step back a bit in case it helps.]

For me a couple use cases mix: I want to conditionalize the progress bar based on whether tty is available. When it's not, seems better to remove certain styles. But it's also the case that in non-tty we really don't want to redraw to screen frequently. But if mechanisms to redraw less frequently are non-performant in some way, or if `indicatif` itself is a potential performance bottleneck`, I would wrap things differently such that I don't use it at all in most situations.

**Comment by mooreniemi** [comment]
Nice, I see [`stdout_with_hz`](https://docs.rs/indicatif/latest/indicatif/struct.ProgressDrawTarget.html#method.stdout_with_hz). But same q: does this actually help, hurt, or have no impact on general performance of indicatif?

**Comment by chris-laplante** [comment]
> Nice, I see [`stdout_with_hz`](https://docs.rs/indicatif/latest/indicatif/struct.ProgressDrawTarget.html#method.stdout_with_hz). But same q: does this actually help, hurt, or have no impact on general performance of indicatif?

To be honest, I'm not sure. But I would sincerely hope that it would help, or at the very least have no impact. I think if using it negatively impacted performance, then that would be a bug in indicatif.

**Comment by djc** [comment]
The draw delta API was removed because there's now implicit rate limiting in indicatif by default. I think when I measured it for the 0.17 release (probably a good idea to read the release notes) it was like a 100x faster in the default configuration. Also progress is now measured with an atomic value there is no more lock contention for the simplest case. In general I think this issue should be closed as the performance issues have been addressed since it was opened.

**Comment by mooreniemi** [comment]
Cool! FWIW I'll leave this benchmark here.

### results

```
read_parquet_without_progress
time:   [1.1644 s 1.1664 s 1.1687 s]

read_parquet_with_progress
time:   [1.2880 s 1.2903 s 1.2929 s]

read_parquet_with_progress_reduced
time:   [1.2975 s 1.2998 s 1.3022 s]
```

Seems like adding Progress does have some slightly impact, but this is hardly scientific. Will update if I dive deeper.


### code

```rust
use criterion::{criterion_group, criterion_main, Criterion};
use indicatif::{ProgressBar, ProgressDrawTarget, ProgressStyle};
use parquet::file::reader::{FileReader, SerializedFileReader};
use parquet::record::Row;
use std::fs::File;

/// File Name: /tmp/reducer.snappy.parquet: 12916 rows
/// Uncompressed Size: 13,141,913
const PARQUET_FILE_PATH: &str = "/tmp/reducer.snappy.parquet";

fn read_parquet_file(path: &str) -> Vec<Row> {
    let file = File::open(path).expect("Failed to open file");
    let reader = SerializedFileReader::new(file).expect("Failed to create reader");
    let iter = reader
        .get_row_iter(None)
        .expect("Failed to get row iterator");

    iter.filter_map(Result::ok).collect::<Vec<Row>>()
}

fn read_parquet_file_with_progress(path: &str) -> Vec<Row> {
    let file = File::open(path).expect("Failed to open file");
    let reader = SerializedFileReader::new(file).expect("Failed to create reader");
    let metadata = reader.metadata();
    let row_group_count = metadata.num_row_groups();

    let pb = ProgressBar::new(row_group_count as u64);
    pb.set_style(
        ProgressStyle::default_bar()
            .template("{msg} [{bar:40.cyan/blue}] {pos:>7}/{len:7} ({percent}%)")
            .expect("valid template")
            .progress_chars("##-"),
    );
    pb.set_message("Reading Parquet");

    let mut rows = Vec::new();
    for row_group_index in 0..row_group_count {
        let row_group_reader = reader
            .get_row_group(row_group_index)
            .expect("Failed to get row group");
        let row_iter = row_group_reader
            .get_row_iter(None)
            .expect("Failed to get row iterator");

        rows.extend(row_iter.filter_map(Result::ok));
        pb.inc(1);
    }

    pb.finish_with_message("Done reading Parquet");

    rows
}

fn read_parquet_file_with_progress_reduced(path: &str) -> Vec<Row> {
    let file = File::open(path).expect("Failed to open file");
    let reader = SerializedFileReader::new(file).expect("Failed to create reader");
    let metadata = reader.metadata();
    let row_group_count = metadata.num_row_groups();

    let pb = ProgressBar::new(row_group_count as u64);
    pb.set_style(
        ProgressStyle::default_bar()
            .template("{msg} [{bar:40.cyan/blue}] {pos:>7}/{len:7} ({percent}%)")
            .expect("valid template")
            .progress_chars("##-"),
    );
    pb.set_message("Reading Parquet");
    pb.set_draw_target(ProgressDrawTarget::stdout_with_hz(1));

    let mut rows = Vec::new();
    for row_group_index in 0..row_group_count {
        let row_group_reader = reader
            .get_row_group(row_group_index)
            .expect("Failed to get row group");
        let row_iter = row_group_reader
            .get_row_iter(None)
            .expect("Failed to get row iterator");

        rows.extend(row_iter.filter_map(Result::ok));
        pb.inc(1);
    }

    pb.finish_with_message("Done reading Parquet");

    rows
}

fn parquet_read_benchmark(c: &mut Criterion) {
    c.bench_function("read_parquet_without_progress", |b| {
        b.iter(|| {
            let rows = read_parquet_file(PARQUET_FILE_PATH);
            criterion::black_box(rows);
        });
    });

    c.bench_function("read_parquet_with_progress", |b| {
        b.iter(|| {
            let rows = read_parquet_file_with_progress(PARQUET_FILE_PATH);
            criterion::black_box(rows);
        });
    });

    c.bench_function("read_parquet_with_progress_reduced", |b| {
        b.iter(|| {
            let rows = read_parquet_file_with_progress_reduced(PARQUET_FILE_PATH);
            criterion::black_box(rows);
        });
    });
}

criterion_group!(benches, parquet_read_benchmark);
criterion_main!(benches);
```

**Comment by djc** [comment]
A 10% impact is definitely more than I would expect, but I won't have time to dig in much deeper. Interested in your results if you do, though!
